---
output:
  md_document:
    variant: markdown_github
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

[![Travis-CI Build Status](https://travis-ci.org/ropensci/EML.svg?branch=master)](https://travis-ci.org/ropensci/EML)
[![codecov.io](https://codecov.io/github/ropensci/EML/coverage.svg?branch=master)](https://codecov.io/github/ropensci/EML?branch=master)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/ropensci/EML?branch=master&svg=true)](https://ci.appveyor.com/project/ropensci/EML)
[![](https://ropensci.org/badges/80_status.svg)](https://github.com/ropensci/onboarding/issues/80)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/EML)](https://cran.r-project.org/package=EML)
[![CRAN RStudio mirror downloads](http://cranlogs.r-pkg.org/badges/EML)](https://CRAN.R-project.org/package=EML)
[![DOI](https://zenodo.org/badge/10894022.svg)](https://zenodo.org/badge/latestdoi/10894022)

<!--
[![Github Stars](https://img.shields.io/github/stars/ropensci/EML.svg?style=social&label=Github)](https://github.com/ropensci/EML)
-->


# EML: The Ecological Metadata Language Standard

EML is a widely used metadata standard in the ecological and environmental sciences.  We strongly recommend that interested users visit the [EML Homepage](https://knb.ecoinformatics.org/#external//emlparser/docs/index.html) for an introduction and thorough documentation of the standard.  Additionally, the scientific article _[The New Bioinformatics: Integrating Ecological Data from the Gene to the Biosphere (Jones et al 2006)](http://doi.org/10.1146/annurev.ecolsys.37.091305.110031)_ provides an excellent introduction into the role EML plays in building metadata-driven data repositories to address the needs of highly hetergenous data that cannot be easily reduced to a traditional vertically integrated database.  At this time, the `EML` R package provides support for the serializing and parsing of all low-level EML concepts, but still assumes some familiarity with the EML standard, particularly for users seeking to create their own EML files.  We hope to add more higher-level functions which will make such familiarity less essential in future development.


*Note*: The `EML` R package is work in progress. Please see the Issues tracker in this repository for details about current issues and development milestonds.

## Installation

You can install the current development version from rOpenSci using:

```{r eval=FALSE}
install.packages("EML", repos = c("http://packages.ropensci.org", "https://cran.rstudio.com"))
```

or install directly from GitHub (after installing the `devtools` package from CRAN):


```{r eval=FALSE}
devtools::install_github("ropensci/EML")
```



## Quickstart

Load the package and read in an EML file:

```{r}
library("EML")
f <- system.file("xsd/test/eml.xml", package = "EML")
eml <- read_eml(f)
```

This creates a native R object called `eml`.  Although this is an "S4 object" type, it uses a "show"
method which displays in an XML-like layout for convenience:

```{r}
eml
```

Validation helps ensure that an EML file can be accurately interpreted by any computer program designed to read EML.  We can validate any EML against the official schema:

```{r, message=TRUE}
# An EML document with no validation errors
eml_validate(eml)

# An EML document with validation errors
invalid_eml <- system.file("xsd/test/example-eml-invalid.xml", package = "EML")

eml_validate(invalid_eml)
```


Write out as EML:

```{r, eval=FALSE}
write_eml(eml, "example.xml")
```


```{r include=FALSE, echo=FALSE}
unlink("example.xml")
```


## Reading EML files

Please see the vignette, [Parsing EML](https://github.com/ropensci/EML/blob/master/vignettes/parsing-EML.Rmd) for a more thorough introduction to reading EML files and extracting relevant metadata.

## Creating EML files

Please see the vignette, [Creating EML](https://github.com/ropensci/EML/blob/master/vignettes/creating-EML.Rmd) for a proper introduction to creating your own EML files. 

## Working with custom units

Users may occassionally need to define custom unit types that EML does not know about.  Please see the vignette, [Working with units](https://github.com/ropensci/EML/blob/master/vignettes/working-with-units.Rmd) for an overview of how units work in EML, how to view the standard unit library, and how to access and define custom units and custom unit types. 

## Archiving, Sharing, and Publishing EML

One of the chief advantages of using EML to manage your own data is the improved ability to search all of your data files, e.g. for the ones that have the particular measurements you need on the particular location or species of interest, and be able to share this data either publicly or securely with select collaborators.  To do so, you will need to upload your EML file and data to a scientific metadata repository that understands EML.  The vignette, [metadata repositories](https://github.com/ropensci/EML/blob/master/vignettes/metadata-repositories.Rmd) describes how to do this.  

## Developer notes

### Creating EML class definitions 

Class definitions (`classes.R`) and methods (`methods.R`) are created programmatically.  From the root of the package, run: `source("inst/create-package/create_package.R")`.  

### Meta

Please note that this project is released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms.
